### PR TITLE
bpo-35210: Use bytes + memoryview + resize instead of bytesarray + array in io.RawIOBase.read

### PR DESCRIPTION
**Pull request by tzickel**
https://bugs.python.org/issue35210

### CODE REVIEW

File: Misc/NEWS.d/next/Core and Builtins/2018-11-10-19-26-00.bpo-35210.nhRfM4.rst [hunk 1]
```diff
@@ -0,0 +1,2 @@
+Use bytes + memoryview + resize instead of bytesarray + bytes in
+io.RawIOBase.read
```
Comment (ZackerySpytz) [inline Misc/NEWS.d/next/Core and Builtins/2018-11-10-19-26-00.bpo-35210.nhRfM4.rst:2]:
```suggestion
:meth:`io.RawIOBase.read`.
```
I think this news entry is a little too vague.

File: Modules/_io/iobase.c [hunk 2]
```diff
@@ -898,36 +898,58 @@ static PyObject *
...
         return NULL;
     }
 
-    res = PyBytes_FromStringAndSize(PyByteArray_AsString(b), n);
-    Py_DECREF(b);
-    return res;
+    if (newn < n) {
+        if (refcnt == 1) {
+            _PyBytes_Resize(&b, newn);
+        } else {
```
Comment (ZackerySpytz) [inline Modules/_io/iobase.c:938]: According to PEP 7, the `}` should be by itself on the previous line (https://www.python.org/dev/peps/pep-0007/#code-lay-out).
```suggestion
        }
        else {
```

File: Modules/_io/iobase.c [hunk 3]
```diff
@@ -898,36 +898,58 @@ static PyObject *
...
+            inside the readinto function */
+            mem = PyBytes_FromStringAndSize(PyBytes_AS_STRING(b), newn);
+            Py_DECREF(b);
+            b = mem;
+        }
+        return b;
+    } else if (newn == n) {
+        return b;
+    } else {
+        PyErr_SetString(PyExc_RuntimeError, "readinto() returned bad value");
```
Comment (ZackerySpytz) [inline Modules/_io/iobase.c:949]: I don't think `RuntimeError` should be used here.

File: Modules/_io/iobase.c [hunk 4]
```diff
@@ -898,36 +898,58 @@ static PyObject *
 _io__RawIOBase_read_impl(PyObject *self, Py_ssize_t n)
 /*[clinic end generated code: output=6cdeb731e3c9f13c input=b6d0dcf6417d1374]*/
 {
-    PyObject *b, *res;
+    PyObject *b, *res, *mem;
+    Py_ssize_t newn, refcnt;
```
Comment (ZackerySpytz) [inline Modules/_io/iobase.c:902]: These new variables don't need to be declared at the top of the block.

**Comment by iritkatriel** [comment]
@tzickel Are you still working on this?

**Pull request closed without merging**